p2p/server.py

```diff
@@ -0,0 +1,147 @@
+import asyncio
+import logging
+import secrets
+
+from typing import (
+    List,
+    Tuple,
+)
+
+from eth_keys import datatypes
+
+from eth_utils import big_endian_to_int
+
+from evm.db.chain import AsyncChainDB
+from p2p.auth import (
+    decode_auth_plain,
+    decode_auth_eip8,
+    HandshakeResponder,
+)
+from p2p.cancel_token import CancelToken
+from p2p.constants import (
+    ENCRYPTED_AUTH_MSG_LEN,
+    HASH_LEN,
+)
+from p2p.discovery import DiscoveryProtocol
+from p2p.ecies import ecdh_agree
+from p2p.exceptions import DecryptionError
+from p2p.kademlia import (
+    Address,
+    Node,
+)
+from p2p.peer import (
+    ETHPeer,
+    PeerPool,
+)
+from p2p.utils import sxor
+
+
+class Server:
+    """Server listening for incoming connections"""
+    logger = logging.getLogger("p2p.server.Server")
+
+    def __init__(self,
+                 privkey: datatypes.PrivateKey,
+                 server_address: Address,
+                 chaindb: AsyncChainDB,
+                 bootstrap_nodes: List[str],
+                 network_id: int
+                 ) -> None:
+        self.cancel_token = CancelToken('Server')
+        self.chaindb = chaindb
+        self.privkey = privkey
+        self.server_address = server_address
+        self.network_id = network_id
+        discovery = DiscoveryProtocol(
+            self.privkey, self.server_address, bootstrap_nodes=bootstrap_nodes)
+        self.peer_pool = PeerPool(ETHPeer, self.chaindb, self.network_id, self.privkey, discovery)
+
+    async def run(self) -> None:
+        self.logger.info("Running server...")
+        factory = asyncio.start_server(
+            self.receive_handshake, host=self.server_address.ip, port=self.server_address.udp_port)
+        asyncio.ensure_future(factory)
```

Comment:
The `start_server()` call returns a `asyncio.Server` instance that we must store as an instance attribute in `p2p.Server` so that we can close it before we exit. Something like:

```
async def run(self) -> None:
    self.server = await asyncio.start_server(...)
    await self.cancel_token.wait()
    await self.stop()
```